File: integrations/android/circlebox-integrations/src/main/kotlin/com/circlebox/integrations/CircleBoxSentryMapper.kt
package com.circlebox.integrations

object CircleBoxSentryMapper {
    fun breadcrumbs(envelope: CircleBoxAdapterEnvelope): List<CircleBoxSentryBreadcrumb> {
        return envelope.events.map { event -> breadcrumb(event, sdk = "android", mode = "export_adapter") }
    }

    fun breadcrumb(
        event: CircleBoxAdapterEvent,
        sdk: String = "android",
        mode: String = "realtime_adapter"
    ): CircleBoxSentryBreadcrumb {
        return CircleBoxSentryBreadcrumb(
            category = "circlebox.${event.type}",
            level = levelFor(event.severity),
            message = event.attrs["message"] ?: event.type,
            timestampUnixMs = event.timestampUnixMs,
            data = linkedMapOf(
                "thread" to event.thread,
                "severity" to event.severity,
                "circlebox_source" to "circlebox",
                "circlebox_mode" to mode,
                "circlebox_sdk" to sdk
            ) + event.attrs
        )
    }

    private fun levelFor(severity: String): String {
        return when (severity) {
            "fatal" -> "fatal"
            "error" -> "error"
            "warn" -> "warning"
            else -> "info"
        }
    }
}

File: integrations/ios/CircleBoxIntegrations/Sources/CircleBoxIntegrations/CircleBoxSentryMapper.swift
import Foundation

public enum CircleBoxSentryMapper {
    public static func breadcrumbs(from envelope: CircleBoxAdapterEnvelope) -> [CircleBoxSentryBreadcrumb] {
        envelope.events.map { event in
            breadcrumb(from: event, sdk: "ios", mode: "export_adapter")
        }
    }

    public static func breadcrumb(
        from event: CircleBoxAdapterEvent,
        sdk: String = "ios",
        mode: String = "realtime_adapter"
    ) -> CircleBoxSentryBreadcrumb {
        CircleBoxSentryBreadcrumb(
            category: "circlebox.\(event.type)",
            level: level(for: event.severity),
            message: event.attrs["message"] ?? event.type,
            timestampUnixMs: event.timestampUnixMs,
            data: [
                "thread": event.thread,
                "severity": event.severity,
                "circlebox_source": "circlebox",
                "circlebox_mode": mode,
                "circlebox_sdk": sdk
            ].merging(event.attrs) { current, _ in current }
        )
    }

    private static func level(for severity: String) -> String {
        switch severity {
        case "fatal":
            return "fatal"
        case "error":
            return "error"
        case "warn":
            return "warning"
        default:
            return "info"
        }
    }
}
